api/utils/keys/piv/pincache.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2025 Gravitational, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package piv
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+
+	"github.com/gravitational/teleport/api/utils/keys/hardwarekey"
+)
+
+// pinCache is a PIN cache that supports consumers with varying required TTLs.
+type pinCache struct {
+	clock clockwork.Clock
+
+	mu sync.Mutex
+	// pin is the cached PIN.
+	pin string
+	// pinSetAt is the time when the cached PIN was set. Used to determine whether
+	// the PIN should be considered expired for a specific caller's provided TTL.
+	pinSetAt time.Time
+	// pinExpiry is the expiration time of the cached PIN.
+	pinExpiry time.Time
+}
+
+// newPINCache returns a new PINCache.
+func newPINCache() *pinCache { //nolint:unused // used in yubikey.go with piv build constraint
+	return &pinCache{
+		clock: clockwork.NewRealClock(),
+	}
+}
+
+// PromptOrGetPIN retrieves the cached PIN if set. Otherwise it prompts for the PIN and caches it.
+func (p *pinCache) PromptOrGetPIN(ctx context.Context, prompt hardwarekey.Prompt, requirement hardwarekey.PINPromptRequirement, keyInfo hardwarekey.ContextualKeyInfo, pinCacheTTL time.Duration) (string, error) {
+	// If the provided ttl is 0, it doesn't support caching, so we just prompt.
+	if pinCacheTTL == 0 {
+		return prompt.AskPIN(ctx, requirement, keyInfo)
+	}
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	if pin := p.getPIN(pinCacheTTL); pin != "" {
+		return pin, nil
+	}
+
+	pin, err := prompt.AskPIN(ctx, requirement, keyInfo)
```

Comment:
Should we add the 1 minute timeout here so an unanswered PIN prompt doesn't prevent all future operations from succeeding?